File: scala/src/test/scala/com/gildedrose/GildedRoseTest.scala
package com.gildedrose

import org.scalatest._

class GildedRoseSpec extends FlatSpec with Matchers {

  behavior of "GildedRose"

  it should "properly update normal item" in new Fixture {
    var items = Array[Item](normal)
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).name should equal("+5 Dexterity Vest")
    app.items(0).sellIn should equal(9)
    app.items(0).quality should equal(19)
  }

  it should "properly update normal item after sellIn degrades" in new Fixture {
    var items = Array[Item](normal)
    val app = new GildedRose(items)
    for (_ <- 1 to 12) { app.updateQuality() }
    app.items(0).name should equal("+5 Dexterity Vest")
    app.items(0).sellIn should equal(-2)
    app.items(0).quality should equal(6)
  }

  it should "properly update Brie" in new Fixture {
    var items = Array[Item](brie)
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).name should equal("Aged Brie")
    app.items(0).sellIn should equal(1)
    app.items(0).quality should equal(1)
  }

  it should "properly update Brie after sellIn degrades" in new Fixture {
    var items = Array[Item](brie)
    val app = new GildedRose(items)
    for (_ <- 1 to 6) { app.updateQuality() }
    app.items(0).name should equal("Aged Brie")
    app.items(0).sellIn should equal(-4)
    app.items(0).quality should equal(10)
  }

  it should "properly update Sulfuras" in new Fixture {
    var items = Array[Item](sulfuras)
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).name should equal("Sulfuras, Hand of Ragnaros")
    app.items(0).sellIn should equal(5)
    app.items(0).quality should equal(80)
  }

  it should "properly update Sulfuras after sellIn degrades" in new Fixture {
    var items = Array[Item](sulfuras)
    val app = new GildedRose(items)
    for (_ <- 1 to 6) { app.updateQuality() }
    app.items(0).name should equal("Sulfuras, Hand of Ragnaros")
    app.items(0).sellIn should equal(5)
    app.items(0).quality should equal(80)
  }

  it should "properly update Backstage pass" in new Fixture {
    var items = Array[Item](backstagePass)
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).name should equal("Backstage passes to a TAFKAL80ETC concert")
    app.items(0).sellIn should equal(11)
    app.items(0).quality should equal(21)
  }

  it should "properly update Backstage pass when  5 < sellin <= 10 " in new Fixture {
    var items = Array[Item](backstagePass)
    val app = new GildedRose(items)
    for (_ <- 1 to 3) { app.updateQuality() }
    app.items(0).name should equal("Backstage passes to a TAFKAL80ETC concert")
    app.items(0).sellIn should equal(9)
    app.items(0).quality should equal(24)
  }

  it should "properly update Backstage pass when  sellin <= 5 " in new Fixture {
    var items = Array[Item](backstagePass)
    val app = new GildedRose(items)
    for (_ <- 1 to 8) { app.updateQuality() }
    app.items(0).name should equal("Backstage passes to a TAFKAL80ETC concert")
    app.items(0).sellIn should equal(4)
    app.items(0).quality should equal(35)
  }

  it should "properly update Backstage pass when sellin <0  " in new Fixture {
    var items = Array[Item](backstagePass)
    val app = new GildedRose(items)
    for (_ <- 1 to 13) { app.updateQuality() }
    app.items(0).name should equal("Backstage passes to a TAFKAL80ETC concert")
    app.items(0).sellIn should equal(-1)
    app.items(0).quality should equal(0)
  }

  it should "properly update Conjured" in new Fixture {
    var items = Array[Item](conjured)
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).name should equal("Conjured")
    app.items(0).sellIn should equal(4)
    app.items(0).quality should equal(28)
  }

  it should "properly update Conjured after sellIn degrades" in new Fixture {
    var items = Array[Item](conjured)
    val app = new GildedRose(items)
    for (_ <- 1 to 6) { app.updateQuality() }
    app.items(0).name should equal("Conjured")
    app.items(0).sellIn should equal(-1)
    app.items(0).quality should equal(16)
  }

  it should "Once the sell by date has passed, Quality degrades twice as fast" in new Fixture {
    var items = Array[Item](normal)
    val app = new GildedRose(items)
    for (_ <- 1 to 10) { app.updateQuality() }
    app.items(0).quality should equal(10)
    app.updateQuality()
    app.items(0).quality should equal(8)
  }

  it should "The Quality of an item is never negative" in new Fixture {
    var items = Array[Item](normal)
    val app = new GildedRose(items)
    for (_ <- 1 to 25) { app.updateQuality() }
    app.items(0).quality should equal(0)
  }

  class Fixture {
    val normal = new Item("+5 Dexterity Vest", 10, 20)
    val brie = new Item("Aged Brie", 2, 0)
    val sulfuras = new Item("Sulfuras, Hand of Ragnaros", 5, 80)
    val backstagePass = new Item("Backstage passes to a TAFKAL80ETC concert", 12, 20)
    val conjured = new Item("Conjured", 5, 30)
  }
}

File: Java/src/test/java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import org.junit.Test;

import java.util.stream.Stream;

import static org.junit.Assert.assertEquals;

public class GildedRoseTest {

    Item normalItem() { return new Item("+5 Dexterity Vest", 10, 20); }
    Item agedBrie() { return new Item("Aged Brie", 2, 0); }
    Item sulfuras() { return new Item("Sulfuras, Hand of Ragnaros", 5, 80); }
    Item backstagePass() { return new Item("Backstage passes to a TAFKAL80ETC concert", 12, 20); }
    Item conjured() { return new Item("Conjured", 5, 30); }

    @Test
    public void updateNormalItemOnce() {
        GildedRose app = updateInventory(normalItem(), 1);
        assertEquals("+5 Dexterity Vest", app.items[0].name);
        assertEquals(9, app.items[0].sellIn);
        assertEquals(19, app.items[0].quality);
    }

    @Test
    public void updateNormalItemAfterSellinDegrades() {
        GildedRose app = updateInventory(normalItem(), 12);
        assertEquals("+5 Dexterity Vest", app.items[0].name);
        assertEquals(-2, app.items[0].sellIn);
        assertEquals(6, app.items[0].quality);
    }

    @Test
    public void updateAgedBrieOnce() {
        GildedRose app = updateInventory(agedBrie(), 1);
        assertEquals("Aged Brie", app.items[0].name);
        assertEquals(1, app.items[0].sellIn);
        assertEquals(1, app.items[0].quality);
    }

    @Test
    public void updateAgedBrieAfterSellinDegrades() {
        GildedRose app = updateInventory(agedBrie(), 6);
        assertEquals("Aged Brie", app.items[0].name);
        assertEquals(-4, app.items[0].sellIn);
        assertEquals(10, app.items[0].quality);
    }

    @Test
    public void updateSulfuras() {
        GildedRose app = updateInventory(sulfuras(), 1);
        assertEquals("Sulfuras, Hand of Ragnaros", app.items[0].name);
        assertEquals(5, app.items[0].sellIn);
        assertEquals(80, app.items[0].quality);
    }

    @Test
    public void updateBackstagePass() {
        GildedRose app = updateInventory(backstagePass(), 1);
        assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        assertEquals(11, app.items[0].sellIn);
        assertEquals(21, app.items[0].quality);
    }

    @Test
    public void updateBackstagePassWhenSellinLessThan10() {
        GildedRose app = updateInventory(backstagePass(), 3);
        assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        assertEquals(9, app.items[0].sellIn);
        assertEquals(24, app.items[0].quality);
    }

    @Test
    public void updateBackstagePassWhenSellinLessThan5() {
        GildedRose app = updateInventory(backstagePass(), 8);
        assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        assertEquals(4, app.items[0].sellIn);
        assertEquals(35, app.items[0].quality);
    }

    @Test
    public void updateBackstagePassWhenAfterConcert() {
        GildedRose app = updateInventory(backstagePass(), 13);
        assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        assertEquals(-1, app.items[0].sellIn);
        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void updateConjured() {
        GildedRose app = updateInventory(conjured(), 1);
        assertEquals("Conjured", app.items[0].name);
        assertEquals(4, app.items[0].sellIn);
        assertEquals(28, app.items[0].quality);
    }

    @Test
    public void updateConjuredAfterSellinDegrades() {
        GildedRose app = updateInventory(conjured(), 6);
        assertEquals("Conjured", app.items[0].name);
        assertEquals(-1, app.items[0].sellIn);
        assertEquals(16, app.items[0].quality);
    }

    @Test
    public void qualityShouldDegradeTwiceAsFastAfterSellinDate() {
        GildedRose app = updateInventory(normalItem(), 10);
        assertEquals(10, app.items[0].quality);
        app.updateQuality();
        assertEquals(8, app.items[0].quality);
    }

    @Test
    public void qualityShouldNeverBeNegative() {
        GildedRose app = updateInventory(normalItem(), 25);
        assertEquals(0, app.items[0].quality);
    }

    private GildedRose updateInventory(Item item, int days) {
        Item[] items = new Item[] { item };
        GildedRose app = new GildedRose(items);
        Stream.iterate(0, i -> i).limit(days).forEach(i -> app.updateQuality() );
        return app;
    }

}
